integrations/cdi/jpa-cdi/src/main/java/io/helidon/integrations/cdi/jpa/DelegatingEntityManager.java

```diff
@@ -0,0 +1,434 @@
+ */
+package io.helidon.integrations.cdi.jpa;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.persistence.EntityGraph;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+import javax.persistence.FlushModeType;
+import javax.persistence.LockModeType;
+import javax.persistence.Query;
+import javax.persistence.StoredProcedureQuery;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaDelete;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.CriteriaUpdate;
+import javax.persistence.metamodel.Metamodel;
+
+/**
+ * A partial {@link EntityManager} implementation that forwards all
+ * calls to an underlying {@link EntityManager}.
+ */
+abstract class DelegatingEntityManager implements EntityManager {
+
+
+    /*
+     * Instance fields.
+     */
+
+
+    /**
+     * The {@link EntityManager} to which all operations will be
+     * forwarded if it is non-{@code null}.
+     *
+     * @see #DelegatingEntityManager(EntityManager)
+     *
+     * @see #delegate()
+     *
+     * @see #acquireDelegate()
+     */
+    private final EntityManager delegate;
+
+
+    /*
+     * Constructors.
+     */
+
+
+    /**
+     * Creates a new {@link DelegatingEntityManager}.
+     *
+     * @see #delegate()
+     */
+    DelegatingEntityManager() {
+        this(null);
+    }
+
+    /**
+     * Creates a new {@link DelegatingEntityManager}.
+     *
+     * @param delegate the {@link EntityManager} to which all
+     * operations may be forwarded (but see the {@link #delegate()}
+     * method); may be {@code null} in which case the {@link
+     * #acquireDelegate()} method will be used to supply a delegate
+     * for each method call
+     *
+     * @see #delegate()
+     *
+     * @see #acquireDelegate()
+     */
+    DelegatingEntityManager(final EntityManager delegate) {
+        super();
+        this.delegate = delegate;
+    }
+
+
+    /*
+     * Instance methods.
+     */
+
+
+    /**
+     * Returns the {@link EntityManager} to which a method invocation
+     * should be forwarded.
+     *
+     * <p>This method never returns {@code null}.</p>
+     *
+     * <p>This method will call the {@link #acquireDelegate()} method
+     * if a {@code null} delegate {@link EntityManager} was
+     * {@linkplain #DelegatingEntityManager(EntityManager) supplied at
+     * construction time}.</p>
+     *
+     * @return an {@link EntityManager}; never {@code null}
+     *
+     * @exception javax.persistence.PersistenceException if an error
+     * occurs
+     *
+     * @see #acquireDelegate()
+     *
+     * @see #DelegatingEntityManager(EntityManager)
+     */
+    protected EntityManager delegate() {
+        final EntityManager returnValue;
+        if (this.delegate == null) {
+            returnValue = this.acquireDelegate();
```

Comment:
I see that the `acquireDelegate` JavaDoc explains that its implementation must be fast because it is invoked each time a delegated method is invoked (if no delegate was passed to the constructor initially). So it seems intentional that we don't invoke `acquireDelegate` once (if there is none) and then cache and reuse that one subsequently. I'm sure there's a good reason; I'm just double-checking to make sure this is what we want.

Reply:
Double-checking is always appreciated.  Yes, the JPA specification basically spells this out.  In many different cases, you are required to hand out an `EntityManager` "shell", proxy, whatever you want to call it—and then on every invocation you have to do things like check to see if you're in a transaction, etc.

You inadvertently raise a good point which is that in one version of this PR I had a couple of use cases where I would pass a delegate to the constructor.  Those cases died off gradually—offhand I can't remember if I still use that constructor/feature at all.  I'll check.

Reply:
(Ah, I do use still use that feature in the `NonTransactionalEntityManager` delegate, whose sole job is to throw `TransactionRequiredException`s from certain methods.)